### ISSUE TITLE
support Tensor type hinting

### ISSUE THREAD

**Issue by wj-Mcat** [PaddlePaddle/Paddle#45979]
## 任务目标

在Python层面实现类型提示，可是目前Paddle暂不支持此功能。

## 需求场景

Python虽然说是动态语言，可是在[PEP 484](https://peps.python.org/pep-0484/) 中已然支持类型注释和提示，其中python3.7+对于类型检查和注释最低版本完美支持type hinting 了。可是目前paddle版本暂不支持此功能。

> there is TODO in [tensor.py](https://github.com/PaddlePaddle/Paddle/blob/e6d397e613f3de987548e1e185281c438f0364ab/python/paddle/tensor/tensor.py#L15)

没有此功能，调用tensor的函数是没有任何提示，这对于开发者而言就像使用记事本在写代码一样。

我之前发了一个[issue](https://github.com/PaddlePaddle/Paddle/issues/33203)来描述此问题，可是没有任何进展，然后我自己写了一个工具：[types-paddle](https://pypi.org/project/types-paddle/) 来解决此问题，而且每天都有稳定的下载量。

## 功能描述

希望能够实现以下功能，并能够实现智能提示：

```python
import paddle
from paddle import Tensor

tensor: Tensor = paddle.randn([2,3])
tensor.sum
```

![](https://user-images.githubusercontent.com/10242208/183284234-a608cf6f-16a0-4e5b-bc76-4e88aa988630.gif)

### 替代实现 Alternatives

可使用stubfiles作为中间解决方案，后续这个排期上了之后可替换掉。

stub-files的原理可见：
* [introduction of stub file](https://mypy.readthedocs.io/en/stable/getting_started.html#stubs-intro)
* [PEP 561 – Distributing and Packaging Type Information](https://peps.python.org/pep-0561/)

## 解决方案

其实只需要添加stub-file即可实现tensor类型注释并在IDE中能够智能提示，这对于编程体验上而来至关重要。

实现方式可参考: https://github.com/wj-Mcat/types-paddle ，或者我来提一个PR fix这个问题也是OK的。

**Comment by SigureMo** [comment]
我之前开发过一个按照 PEP 561 所描述 stub only package https://github.com/cattidea/paddlepaddle-stubs ，也在前两天的 [PFCC 第十次会议](https://github.com/PaddlePaddle/community/blob/master/pfcc/2022-09-08-meeting-minutes.md)中提到了这个

Python3.7 是对 Type Hints 非常友好的版本（目前 Python3.6 已经 EOL 了，所以现在是一个很好的时机），希望 Paddle 将来会在内部直接集成 Type Hints

paddlepaddle-stubs 即是你所说的 stubfiles 解决方案，在该解决方案中可先尝试一些类型的写法，且在运行时没有任何副作用，如果将来 Paddle 能够在内部使用 Type Hints 的话应该也会有所受益。

**Comment by wj-Mcat** [comment]
Type Hint对于动态语言太重要了，没有智能提示的日子太难受了，所以希望官方能够关注一下这个问题。

**Comment by gglin001** [comment]
个人觉得不止 `Tensor type hinting`, 如果能够 对 core*.so 生成 stubs, 也是对提高开发体验也是有益的

**Comment by wj-Mcat** [comment]
@SigureMo 你的这个方法改动太大，而且IDE的所有类型提示都会有stub file package，所以你的这个package的维护工作量将会很大。

此外，F12会走到你的pyi 文件里面来，脱离了源码。

个人感觉有这两个弊端。

**Comment by wj-Mcat** [comment]
> 个人觉得不止 `Tensor type hinting`, 如果能够 对 core*.so 生成 stubs, 也是对提高开发体验也是有益的

这个就是生成对应的头文件就行，应该有对应的实现……………………吧？

**Comment by SigureMo** [comment]
> 个人觉得不止 Tensor type hinting, 如果能够 对 core*.so 生成 stubs, 也是对提高开发体验也是有益的

这个的话 paddle._C_ops 也许可以考虑同样直接通过 YAML 文件直接生成？一些其他的 pybind 的 API 可能要手写？

**Comment by SigureMo** [comment]
> @SigureMo 你的这个方法改动太大，而且IDE的所有类型提示都会有stub file package，所以你的这个package的维护工作量将会很大。

维护量大是肯定的，如果想在 Paddle 内部集成的话比这麻烦的多

> 此外，F12会走到你的pyi 文件里面来，脱离了源码。

除去没有「实现」的纯 pyi 文件，只要有对应「实现」的 py 文件的 API 都会跳转到 py 文件，这应该没啥问题，Tensor 这种没有 py「实现」的肯定会跳转到 pyi 文件

**Comment by gglin001** [comment]
> 这个的话 paddle._C_ops 也许可以考虑同样直接通过 YAML 文件直接生成？一些其他的 pybind 的 API 可能要手写？

是的, paddle._C_ops 可以通过 yaml 生成, 其他的大部分的api 应该也是有可能自动生成的, 比如使用 https://pypi.org/project/clang/ 分析对应的 cpp

**Comment by gglin001** [comment]
也许可以在 paddle 里面添加一份基础的/有普遍需求的 一些 pyi, 比如这里的 Tensor api, 不一定是全部都放进来?

**Comment by wj-Mcat** [comment]
> 也许可以在 paddle 里面添加一份基础的/有普遍需求的 一些 pyi, 比如这里的 Tensor api, 不一定是全部都放进来?

赞成，tensor目前是最高优，而且所有都用pyi 来添加代价很大，而且治标不治本: in-line annotation.

**Comment by SigureMo** [comment]
> 赞成，tensor目前是最高优，而且**所有都用pyi** 来添加代价很大，而且治标不治本: in-line annotation.

PEP 561 的包的意义不是将来直接把所有 pyi 文件直接 copy 到 Paddle 内部，而是将 C/C++ 实现的部分的 pyi copy 进去，其余部分已有 Python 实现的这些 pyi 只能作为参考，还是需要集成到 inline 的 Type Hints 的，而不是在 py 实现旁边放一个 pyi 文件

> 也许可以在 paddle 里面添加一份基础的/有普遍需求的 一些 pyi, 比如这里的 Tensor api, 不一定是全部都放进来?

确实可以，应该直接将 tensor.pyi copy 过去后稍微调整下即可，关于 paddle._C_ops 可以考虑在生成 C++ 代码时顺便生成一份 pyi，其余也可以考虑扫描后慢慢添加，不过也许在 PEP 561 包中先进行这些实验是更好的选择，因为 Paddle 目前貌似还需要支持 Python3.6……

**Comment by wj-Mcat** [comment]
> PEP 561 的包的意义不是将来直接把所有 pyi 文件直接 copy 到 Paddle 内部，而是将 C/C++ 实现的部分的 pyi copy 进去

个人看法 😂 ：[PEP 561](https://peps.python.org/pep-0561/#specification) 里面没有提到任何关于C/C++部分，核心是围绕：`type information`，所以只要能够添加类型信息让IDE以及typechecker(eg: mypy)能够获取到类型信息即可.  在tensor.py旁边放一个tensor.pyi就是为了最小程度上添加最高优的 **tensor** `type information`.

**Comment by SigureMo** [comment]
> 个人看法 😂 ：[PEP 561](https://peps.python.org/pep-0561/#specification) 里面没有提到任何关于C/C++部分，核心是围绕：type information，所以只要能够添加类型信息让IDE以及typechecker(eg: mypy)能够获取到类型信息即可. 在tensor.py旁边放一个tensor.pyi就是为了最小程度上添加最高优的 tensor type information.

抱歉，这里是我表述错了，应该是 PEP 561 stub only 包

> 不过也许在 PEP 561 包中先进行这些实验是更好的选择，因为 Paddle 目前貌似还需要支持 Python3.6……

如果只添加 Tensor 的话也确实不需要考虑这个问题，因为 Tensor 大概率也不需要 Python3.7 及之后才支持的特性，这个也许可以考虑尝试先添加下

**Comment by wj-Mcat** [comment]
期待这个功能早日完成，然后下游套件的type hint也都可以开始支持起来了。

**Comment by wj-Mcat** [comment]
请问一下这个feature是否有任何进展？可link此讨论，这样订阅此issue的开发者都可收到最新进展。

非常期待此feature的最新相关进展。

**Comment by gglin001** [comment]
似乎没看到有啥进展? 
@wj-Mcat 感觉不如 按照你的原始提议

> 实现方式可参考: https://github.com/wj-Mcat/types-paddle ，或者我来提一个PR fix这个问题也是OK的

先提交一个pr...

**Comment by wj-Mcat** [comment]
@gglin001 我觉得可，而且可以优化成省去install的方式。就是不知道paddle组委会的人怎么看这件事，

**Comment by jzhang533** [comment]
> 就是不知道paddle组委会的人怎么看这件事

我们并没有paddle组委会这么个组织，从内部的技术会议上，目前的安排看，还没有内部的人力可以做这件事情。
如果社区里有人愿意做这件事情的话，我可以帮忙安排一次评审会议，征求一下意见和想法，评审通过后，就可以开始干了。

**Comment by jzhang533** [comment]
> 因为 Paddle 目前貌似还需要支持 Python3.6……
develop 分支已经去掉对python 3.6的支持了 #46477。

**Comment by wj-Mcat** [comment]
> > 就是不知道paddle组委会的人怎么看这件事
> 
> 我们并没有paddle组委会这么个组织，从内部的技术会议上，目前的安排看，还没有内部的人力可以做这件事情。 如果社区里有人愿意做这件事情的话，我可以帮忙安排一次评审会议，征求一下意见和想法，评审通过后，就可以开始干了。

可
